src/lib/vasi-sync/src/lazy_lock.rs

```diff
@@ -0,0 +1,176 @@
+        loop {
+            match state {
+                InitState::Uninitd => {
+                    match self.init_state.compare_exchange(
+                        InitState::Uninitd.into(),
+                        InitState::Initializing.into(),
+                        atomic::Ordering::Acquire,
+                        atomic::Ordering::Relaxed,
+                    ) {
+                        Ok(_) => {
+                            // We have the initialization lock.
+                            self.value.get_mut().with(|p: *mut MaybeUninit<T>| unsafe {
+                                let initializer =
+                                    self.initializer.get_mut().deref().take().unwrap();
+                                (*p).write(initializer.initialize());
+                            });
+                            state = InitState::Initd;
+                            self.init_state
+                                .store(state.into(), atomic::Ordering::Release);
+                            break;
+                        }
+                        Err(v) => {
+                            // It changed out from under us; update and try again.
+                            state = v.try_into().unwrap();
+                        }
+                    }
+                }
+                InitState::Initializing => {
+                    // Another thread is initializing. Spin.
+                    // TODO: block?
+                    rustix::process::sched_yield();
+                    state = InitState::try_from(self.init_state.load(atomic::Ordering::Acquire))
+                        .unwrap();
+                }
+                InitState::Initd => {
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Force initialization and return a reference object.
+    ///
+    /// std's LazyLock returns an actual reference here and implements Deref.
+    /// We can't because of our usage of loom's UnsafeCell, which doesn't support
+    /// getting a raw reference that's not borrowed from an intermediate temp object.
+    #[inline]
+    pub fn force(&self) -> Ref<T, Init> {
+        let state = InitState::try_from(self.init_state.load(atomic::Ordering::Acquire)).unwrap();
+        if state != InitState::Initd {
+            self.init(state);
+        }
+        Ref {
+            val: self.value.get(),
+            _phantom: PhantomData,
+        }
+    }
+
+    #[inline]
+    pub fn initd(&self) -> bool {
+        InitState::try_from(self.init_state.load(atomic::Ordering::Acquire)).unwrap()
+            == InitState::Initd
+    }
+}
+
+impl<T, Init> Drop for LazyLock<T, Init> {
+    fn drop(&mut self) {
+        if InitState::try_from(self.init_state.load(atomic::Ordering::Acquire)).unwrap()
+            == InitState::Initd
+        {
+            unsafe { self.value.get_mut().deref().assume_init_drop() }
+        }
+    }
+}
+
+pub struct Ref<'a, T, Init> {
```

Comment:
Derive `Clone` (or `Copy` if `ConstPtr` allows it)?

Reply:
Can't; `ConstPtr` implements neither. https://docs.rs/loom/latest/loom/cell/struct.ConstPtr.html